### PR TITLE
py/emitinlinethumb: Use 16 bit encodings for PUSH LR / POP PC

### PR DESCRIPTION
**Pull request by czietz**
The Thumb instruction set has special 16 bit encodings for PUSH involving LR and POP involving PC, which are commonly used in nested functions. Using this encoding is particularly important for ARMv6-M, where the more general 32 bit encoding of PUSH and POP is unavailable.

Without this patch, on a CPU implementing ARMv6-M (such as the Raspberry Pi Pico), the following code:
```python
@micropython.asm_thumb
def minimal_example():
    push({r0,lr})
    # [...]
```
... would fail with _SyntaxError: unsupported Thumb instruction 'push' with 1 arguments_, even though this is a supported instruction on ARMv6-M.

Reference for instruction set encoding:
arm-push
arm-pop

Note that both ARMv6-M and ARMv7-M support this encoding:
* ARMv6-M Architecture Reference Manual
https://developer.arm.com/documentation/ddi0419/latest/
* ARMv7-M Architecture Reference Manual
https://developer.arm.com/documentation/ddi0403/latest

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for the contribution, this is a nice improvement!

Can you please add a test to `tests/inlineasm/asmpushpop.py`?  Just another function that calls push/pop with lr.

**Comment by czietz** [comment]
> Can you please add a test to tests/inlineasm/asmpushpop.py? Just another function that calls push/pop with lr.

Done. Note that ```pop {pc}``` inevitably returns, so I had to wrap it into a subroutine.

I also fixed the code and commit formatting issues found by the respective workflows.

File: tests/inlineasm/asmpushpop.py [hunk 1]
```diff
@@ -5,5 +5,15 @@ def f(r0, r1, r2):
...
+    label(SUBROUTINE)
+    push({lr})  # push return address
+    mov(r0, 7)
+    pop({pc})   # return
+    label(START)
+    bl(SUBROUTINE)
+
 
-print(f(0, 1, 2))
+print(f(0, 1, 2), g())
```
Comment (dpgeorge) [inline tests/inlineasm/asmpushpop.py:19]: please write the new test as a separate print, ie:
```python
print(f(...))
print(g())
```

Comment (czietz) [reply tests/inlineasm/asmpushpop.py]: Done. See new commit 3037b57e8b86d54f4c6ac8b4a6353bc1f025c391.

**Comment by czietz** [comment]
Commit 5c0cba32c80c5c05c24dfca6468f8cc0ab4f86b7 fixes the source code formatting as indicated by this Black run:
https://github.com/micropython/micropython/runs/5961454308?check_suite_focus=true

**Review by dpgeorge — APPROVED** [review]

**Comment by dpgeorge** [comment]
Thanks for updating.  Rebased and merged in b0bcb3862b95d54b0457d61d779826641e5784fd and d6c59c9d728854ecef4c284cc7cb2428021ea5d4

**Pull request closed without merging**